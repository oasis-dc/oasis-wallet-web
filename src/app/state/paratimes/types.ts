import { StringifiedBigInt } from 'types/StringifiedBigInt'
import { ErrorPayload } from 'types/errors'
import { ParaTime } from '../../../config'

export enum TransactionFormSteps {
  TransferType = 'transferType',
  ParaTimeSelection = 'paraTimeSelection',
  TransactionRecipient = 'transactionRecipient',
  TransactionAmount = 'transactionAmount',
  TransactionConfirmation = 'transactionConfirmation',
  TransactionSummary = 'transactionSummary',
  TransactionError = 'transactionError',
}

export enum TransactionTypes {
  Deposit = 'deposit',
  Withdraw = 'withdraw',
}

export interface TransactionForm {
  amount: string
  confirmTransfer: boolean
  confirmTransferToValidator: boolean
  confirmTransferToForeignAccount: boolean
  defaultFeeAmount: string
  // compatible with oasisRT.signatureSecp256k1
  ethPrivateKey: string
  // provided by user and used in form inputs allowing back and forth form navigation
  ethPrivateKeyRaw: string
  feeAmount: string
  feeGas: string
  paraTime?: ParaTime
  recipient: string
  type: TransactionTypes | undefined
}

export interface ParaTimesState {
  balance: StringifiedBigInt
  isLoading: boolean
  transactionError?: ErrorPayload
  transactionForm: TransactionForm
  transactionFormStep: TransactionFormSteps
}

export type Runtime = {
  address: string
  id: string
  decimals: number
  feeGas: bigint
}

export type ParaTimeTransaction = Pick<
  TransactionForm,
  'amount' | 'ethPrivateKey' | 'feeAmount' | 'feeGas' | 'recipient' | 'type'
>
