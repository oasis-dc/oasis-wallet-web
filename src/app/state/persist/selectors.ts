import { createSelector } from '@reduxjs/toolkit'

import { RootState } from 'types'
import { getInitialState } from '.'

export const selectSlice = (state: RootState) => state.persist || getInitialState()

export const selectLoading = createSelector([selectSlice], state => state.loading)
export const selectEnteredWrongPassword = createSelector([selectSlice], state => state.enteredWrongPassword)
export const selectNeedsPassword = createSelector(
  [selectSlice],
  state => state.hasPersistedProfiles && !state.stringifiedEncryptionKey,
)
export const selectHasV0StorageToMigrate = createSelector([selectSlice], state => state.hasV0StorageToMigrate)
export const selectSkipUnlockingOnInit = createSelector(
  [selectSlice],
  state => !state.hasPersistedProfiles && !state.stringifiedEncryptionKey,
)
export const selectIsPersistenceUnsupported = createSelector(
  [selectSlice],
  state => state.isPersistenceUnsupported,
)

export const selectStringifiedEncryptionKey = createSelector(
  [selectSlice],
  state => state.stringifiedEncryptionKey,
)
