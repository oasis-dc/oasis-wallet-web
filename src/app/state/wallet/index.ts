import { PayloadAction, current } from '@reduxjs/toolkit'
import { createSlice } from 'utils/@reduxjs/toolkit'

import {
  BalanceUpdatePayload,
  OpenFromPrivateKeyPayload,
  OpenSelectedAccountsPayload,
  Wallet,
  WalletState,
} from './types'

export const initialState: WalletState = {
  wallets: {},
  selectedWallet: undefined,
}

export const walletSlice = createSlice({
  name: 'wallet',
  initialState,
  reducers: {
    openWalletFromMnemonic(state, action: PayloadAction<OpenSelectedAccountsPayload>) {},
    openWalletFromPrivateKey(state, action: PayloadAction<OpenFromPrivateKeyPayload>) {},
    openWalletsFromLedger(state, action: PayloadAction<OpenSelectedAccountsPayload>) {},
    deleteWallet(state, action: PayloadAction<string>) {
      delete state.wallets[action.payload]
    },
    selectFirstWallet(state) {
      state.selectedWallet = Object.values(current(state.wallets))[0].address
    },
    selectWallet(state, action: PayloadAction<string | undefined>) {
      state.selectedWallet = action.payload
    },
    fetchWallet(state, action: PayloadAction<Wallet>) {},
    updateBalance(state, action: PayloadAction<BalanceUpdatePayload>) {
      // We simply ignore the updateBalance action in rare cases when opening a
      // second tab while account is loading:
      // - tab2 initializes
      // - tab1 dispatches updateBalance
      // - tab2 asks to sync state
      // - tab2 receives updateBalance for an account it doesn't have
      // - tab1 responds with state
      // - tab2 receives synced state with the account it didn't have
      //
      // Alternatives:
      // - Remove updateBalance from whitelistTabSyncActions and have an
      //   outdated balance in case tab2 makes a transaction.
      // - Refactor redux-state-sync to ignore actions until tab2 receives
      //   synced state. But tab1 needs to listen to actions too, even tho it
      //   never asks for synced state.
      if (state.wallets[action.payload.address]) {
        state.wallets[action.payload.address].balance = {
          ...state.wallets[action.payload.address].balance,
          ...action.payload.balance,
        }
      }
    },
    walletOpened(state, action: PayloadAction<Wallet>) {
      const newWallet = action.payload
      state.wallets[newWallet.address] = newWallet
      state.selectedWallet ??= newWallet.address
    },
    setWalletName(state, action: PayloadAction<{ address: string; name: string }>) {
      state.wallets[action.payload.address].name = action.payload.name
    },
  },
})

export const { actions: walletActions } = walletSlice
