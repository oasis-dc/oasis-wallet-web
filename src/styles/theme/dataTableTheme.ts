import { IDataTableStyles, ITheme } from 'react-data-table-component'

export const dataTableTheme = {
  background: { default: 'unset' },
  highlightOnHover: {
    default: '#88888833',
    text: 'unset',
  },
  text: {
    primary: 'unset',
    secondary: 'unset',
    disabled: 'unset',
  },
  sortFocus: {
    default: 'unset',
  },
  divider: { default: '#AAAAAAaa' },
} as Partial<ITheme>

export const dataTableCustomStyles: IDataTableStyles = {
  headCells: {
    style: {
      fontSize: '16px',
    },
  },
  rows: {
    style: {
      fontSize: '16px',
    },
  },
}
