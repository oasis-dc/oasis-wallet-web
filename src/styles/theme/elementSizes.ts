export const sidebarSmallSizeLogo = 32
export const sidebarMediumSizeLogo = 40
export const sidebarLargeSizeLogo = 40
export const mobileToolbarZIndex = 3
export const mobileToolbarButtonsZIndex = mobileToolbarZIndex + 1
export const buildBannerZIndex = mobileToolbarZIndex - 1

export const mobileFooterNavigationHeight = '4rem'
