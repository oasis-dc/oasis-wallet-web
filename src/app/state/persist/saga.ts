import { AnyAction } from '@reduxjs/toolkit'
import { actionChannel, call, fork, put, select, take } from 'typed-redux-saga'
import { isActionSynced } from 'redux-state-sync'
import { persistActions, STORAGE_FIELD } from './index'
import {
  base64andStringify,
  decryptWithKey,
  decryptWithPassword,
  deriveKeyFromPassword,
  encryptWithKey,
  fromBase64andParse,
} from './encryption'
import { RootState } from 'types'
import {
  EncryptedString,
  KeyWithSalt,
  PersistState,
  PersistedRootState,
  SetUnlockedRootStatePayload,
} from './types'
import { PasswordWrongError } from 'types/errors'
import { walletActions } from 'app/state/wallet'
import { selectUnlockedStatus } from 'app/state/selectUnlockedStatus'
import { runtimeIs } from 'config'
import { backupAndDeleteV0ExtProfile, readStorageV0 } from '../../../utils/walletExtensionV0'
import { selectStringifiedEncryptionKey } from './selectors'

function* watchPersistAsync() {
  yield* fork(function* () {
    const channelQueue = yield* actionChannel<AnyAction>('*')
    while (true) {
      // TODO: could flush(channelQueue) to speedup multiple consecutive writes
      const action: AnyAction = yield* take(channelQueue)
      // Make queue of async operations, without forking, to avoid race
      // conditions (e.g. multiple writes to localStorage in wrong order).
      // Problems with queue:
      // - select() returns latest state instead of state when action was dispatched.
      //   I think that still works correctly.
      yield* call(handleAsyncPersistActions, action)
    }
  })
}

function* handleAsyncPersistActions(action: AnyAction) {
  if (persistActions.setPasswordAsync.match(action)) {
    yield* call(setPasswordAsync, action.payload.password)
  } else if (persistActions.unlockAsync.match(action)) {
    yield* call(unlockAsync, action)
  } else if (persistActions.lockAsync.match(action)) {
    yield* call(lockAsync, action)
  } else if (persistActions.deleteProfileAsync.match(action)) {
    yield* call(deleteProfileAsync, action)
  } else if (persistActions.updatePasswordAsync.match(action)) {
    yield* call(updatePasswordAsync, action)
  } else if (persistActions.setUnlockedRootState.match(action)) {
    // Skip encrypting the same state
  } else if (persistActions.resetRootState.match(action)) {
    // Skip encrypting the empty state
    yield* call(resetRootState, action)
  } else if (persistActions.setWrongPassword.match(action)) {
    // Skip encrypting the same state
  } else if (persistActions.finishV0Migration.match(action)) {
    yield* call(finishV0Migration, action)
  } else {
    // Encrypt after other actions

    yield* call(encryptAndPersistState, action)
  }
}

// Handlers
function* setPasswordAsync(password: string) {
  /**
   * Latest state does not match state immediately after action was dispatched,
   * but when it is queued by {@link watchPersistAsync}.
   */
  const latestState: RootState = yield* select()
  if (latestState.persist.isPersistenceUnsupported) throw new Error('Persistence not supported')
  const keyWithSalt = yield* call(deriveKeyFromPassword, password)
  const encryptedState = yield* call(encryptState, latestState, keyWithSalt)
  window.localStorage.setItem(STORAGE_FIELD, encryptedState)
  const encryptedState2 = window.localStorage.getItem(
    STORAGE_FIELD,
  ) as EncryptedString<PersistedRootState> | null
  const unlockedPayload = yield* call(decryptState, encryptedState2!, password)
  yield* put(persistActions.setUnlockedRootState(unlockedPayload))
}

function* unlockAsync(action: ReturnType<typeof persistActions.unlockAsync>) {
  const encryptedState = window.localStorage.getItem(
    STORAGE_FIELD,
  ) as EncryptedString<PersistedRootState> | null
  if (!encryptedState) throw new Error('Unexpected unlock action while no state is locked')
  try {
    const unlockedPayload = yield* call(decryptState, encryptedState, action.payload.password)
    yield* put(persistActions.setUnlockedRootState(unlockedPayload))
  } catch (error) {
    if (error instanceof PasswordWrongError) {
      yield* put(persistActions.setWrongPassword())
    } else {
      throw error
    }
  }
}

function* updatePasswordAsync(action: ReturnType<typeof persistActions.updatePasswordAsync>) {
  const encryptedState = window.localStorage.getItem(
    STORAGE_FIELD,
  ) as EncryptedString<PersistedRootState> | null
  if (!encryptedState) throw new Error('Unexpected update password action while no state is locked')
  try {
    // used only for a current password validation, doesn't use decrypted result
    yield* call(decryptState, encryptedState, action.payload.currentPassword)
    yield* put(persistActions.setPasswordAsync({ password: action.payload.password }))
  } catch (error) {
    if (error instanceof PasswordWrongError) {
      yield* put(persistActions.setWrongPassword())
    } else {
      throw error
    }
  }
}

function* lockAsync(action: ReturnType<typeof persistActions.lockAsync>) {
  yield* put(persistActions.resetRootState())
  // Implies state.loading = false
}

function* deleteProfileAsync(action: ReturnType<typeof persistActions.deleteProfileAsync>) {
  yield* call([window.localStorage, window.localStorage.removeItem], STORAGE_FIELD)
  yield* put(persistActions.resetRootState())
  // Implies state.loading = false
}

function* resetRootState(action: ReturnType<typeof persistActions.resetRootState>) {
  const unlockedStatus = yield* select(selectUnlockedStatus)
  // Redirect home to prevent infinite loading if user closes unpersisted wallet in another tab.
  if (isActionSynced(action) && unlockedStatus === 'emptyUnpersisted') {
    // Note: requires saga to be running in popup in extension, not in a service worker.
    yield* call([window.location, window.location.assign], '/')
  }
}

function* finishV0Migration(action: ReturnType<typeof persistActions.finishV0Migration>) {
  yield* put(
    persistActions.setUnlockedRootState({
      persistedRootState: action.payload.persistedRootState,
      stringifiedEncryptionKey: 'skipped',
    }),
  )
  yield* call(setPasswordAsync, action.payload.password)

  // Just being extra careful before deleting old state
  const stateAfterMigration: RootState = yield* select()
  if (
    Object.values(action.payload.persistedRootState.wallet.wallets).length !==
    Object.values(stateAfterMigration.wallet.wallets).length
  ) {
    throw new Error('Something went wrong during V0 state migration')
  }

  yield* call(backupAndDeleteV0ExtProfile)
}

// Encrypting
async function encryptState(
  state: RootState,
  keyWithSalt: KeyWithSalt,
): Promise<EncryptedString<PersistedRootState>> {
  const persistedRootState: PersistedRootState = {
    contacts: state.contacts,
    evmAccounts: state.evmAccounts,
    theme: state.theme,
    wallet: state.wallet,
    network: state.network,
  }
  const encryptedState = await encryptWithKey(keyWithSalt, persistedRootState)
  return encryptedState
}

async function decryptState(
  encryptedState: EncryptedString<PersistedRootState>,
  password: string,
): Promise<SetUnlockedRootStatePayload> {
  const persistedRootState: PersistedRootState = await decryptWithPassword(password, encryptedState)
  const keyWithSalt = await deriveKeyFromPassword(password)
  const stringifiedEncryptionKey = base64andStringify(keyWithSalt)
  return { persistedRootState, stringifiedEncryptionKey }
}

function* encryptAndPersistState(action: AnyAction) {
  /**
   * Latest state does not match state immediately after action was dispatched,
   * but when it is queued by {@link watchPersistAsync}.
   */
  const latestState: RootState = yield* select()
  if (latestState.persist.stringifiedEncryptionKey === 'skipped') return
  if (action.type.startsWith('@')) return // Ignore @@INIT from redux-devtools-instrument
  if (action.type.startsWith('&')) return // Ignore e.g. &_GET_INIT_STATE from redux-state-sync
  if (isActionSynced(action)) return // Ignore actions synced across tabs from redux-state-sync

  // Handle actions while state is locked
  // TODO: If we used private/public key-pair we could encrypt and persist
  // state while state is locked. But not if we support multiple profiles.
  if (!latestState.persist.stringifiedEncryptionKey) {
    if (walletActions.walletOpened.match(action)) {
      throw new Error('Could not add account while state is locked.')
    } else {
      return // Ignore non-essential actions emitted while state is locked.
    }
  }

  const keyWithSalt: KeyWithSalt = fromBase64andParse(latestState.persist.stringifiedEncryptionKey)
  const encryptedState = yield* call(encryptState, latestState, keyWithSalt)
  window.localStorage.setItem(STORAGE_FIELD, encryptedState)
}

function* retainEncryptionKeyBetweenPopupReopenings() {
  yield* fork(function* () {
    if (runtimeIs !== 'extension') return

    // Read before rewriting.
    const encryptedState = window.localStorage.getItem(
      STORAGE_FIELD,
    ) as EncryptedString<PersistedRootState> | null
    if (encryptedState) {
      try {
        const stringifiedEncryptionKey = yield* call(readSharedExtMemory)
        if (!stringifiedEncryptionKey) throw new Error('skip')
        if (stringifiedEncryptionKey === 'skipped') throw new Error('skip')
        const keyWithSalt: KeyWithSalt = fromBase64andParse(stringifiedEncryptionKey)
        const persistedRootState = yield* call(
          decryptWithKey<PersistedRootState>,
          keyWithSalt,
          encryptedState,
        )
        yield* put(persistActions.setUnlockedRootState({ persistedRootState, stringifiedEncryptionKey }))
      } catch (error) {
        // Ignore
      }
    }

    const channelQueue = yield* actionChannel<AnyAction>('*')
    let previousStringifiedEncryptionKey: PersistState['stringifiedEncryptionKey'] = undefined
    while (true) {
      yield* take(channelQueue)
      const stringifiedEncryptionKey = yield* select(selectStringifiedEncryptionKey)
      if (stringifiedEncryptionKey !== previousStringifiedEncryptionKey) {
        previousStringifiedEncryptionKey = stringifiedEncryptionKey
        yield* call(writeSharedExtMemory, stringifiedEncryptionKey)
      }
    }
  })
}

async function writeSharedExtMemory(stringifiedEncryptionKey: PersistState['stringifiedEncryptionKey']) {
  if (runtimeIs !== 'extension') return
  const browser = await import('webextension-polyfill')
  // Chrome ignores undefined values https://github.com/w3c/webextensions/issues/263
  if (stringifiedEncryptionKey) {
    await browser.storage.session.set({ stringifiedEncryptionKey })
  } else {
    await browser.storage.session.remove('stringifiedEncryptionKey')
  }
}

async function readSharedExtMemory() {
  if (runtimeIs !== 'extension') return
  const browser = await import('webextension-polyfill')
  const storage = await browser.storage.session.get('stringifiedEncryptionKey')
  return storage.stringifiedEncryptionKey as PersistState['stringifiedEncryptionKey']
}

export function* persistSaga() {
  yield* watchPersistAsync()
  yield* retainEncryptionKeyBetweenPopupReopenings()
  const storageV0 = yield* call(readStorageV0)
  yield* put(persistActions.setHasV0StorageToMigrate(!!storageV0?.chromeStorageLocal.keyringData))
}
