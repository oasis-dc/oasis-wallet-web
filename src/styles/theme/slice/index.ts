import { PayloadAction } from '@reduxjs/toolkit'
import { createSlice } from 'utils/@reduxjs/toolkit'

import { getThemeFromStorage, saveTheme } from '../utils'
import { ThemeState } from './types'

export const getInitialState = (): ThemeState => ({
  selected: getThemeFromStorage() || 'system',
})

export const themeSlice = createSlice({
  name: 'theme',
  initialState: () => getInitialState(),
  reducers: {
    changeTheme(state, action: PayloadAction<'dark' | 'light' | 'system'>) {
      saveTheme(action.payload)
      state.selected = action.payload
    },
  },
})

export const { actions: themeActions } = themeSlice
