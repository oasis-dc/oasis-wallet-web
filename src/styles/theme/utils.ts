import { ThemeKeyType } from './slice/types'

/* istanbul ignore next line */
export const isSystemDark = window?.matchMedia
  ? window.matchMedia('(prefers-color-scheme: dark)')?.matches
  : undefined

export const getTargetTheme = (theme: 'light' | 'dark' | 'system') =>
  theme === 'system' ? (isSystemDark ? 'dark' : 'light') : theme

export function saveTheme(theme: ThemeKeyType) {
  window.localStorage.setItem('selectedTheme', theme)
}

/* istanbul ignore next line */
export function getThemeFromStorage(): ThemeKeyType | null {
  return window.localStorage.getItem('selectedTheme') as ThemeKeyType
}
