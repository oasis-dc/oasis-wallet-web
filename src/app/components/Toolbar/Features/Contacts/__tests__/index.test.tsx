import { render, screen } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import userEvent from '@testing-library/user-event'
import { Provider } from 'react-redux'
import { configureAppStore } from 'store/configureStore'
import { ThemeProvider } from 'styles/theme/ThemeProvider'
import { Wallet } from 'app/state/wallet/types'
import { PersistState } from 'app/state/persist/types'
import { Contacts } from '../'

// needed to test Trans component
jest.unmock('react-i18next')

const renderComponent = (store: any) =>
  render(
    <Provider store={store}>
      <MemoryRouter>
        <ThemeProvider>
          <Contacts closeHandler={() => {}} />
        </ThemeProvider>
      </MemoryRouter>
    </Provider>,
  )

describe('<Contacts  />', () => {
  let store: ReturnType<typeof configureAppStore>

  const unlockedProfile = {
    wallet: {
      selectedWallet: 'dummy',
      wallets: {
        dummy: {
          address: 'dummy',
        } as Wallet,
      },
    },
    persist: {
      hasPersistedProfiles: true,
      stringifiedEncryptionKey: 'unlockedProfile',
    } as PersistState,
  }

  beforeEach(() => {
    store = configureAppStore(unlockedProfile)
  })

  it('should render unavailable state', () => {
    renderComponent(configureAppStore({}))
    expect(screen.getByText(/To start adding contacts create a profile while/)).toBeInTheDocument()
    expect(screen.queryByText('Add contact')).not.toBeInTheDocument()
    expect(screen.getByRole('button', { name: 'opening a wallet' })).toBeInTheDocument()
  })

  it('should match snapshot', () => {
    const storeWithContacts = configureAppStore({
      ...unlockedProfile,
      contacts: {
        oasis1qq2vzcvxn0js5unsch5me2xz4kr43vcasv0d5eq4: {
          address: 'oasis1qq2vzcvxn0js5unsch5me2xz4kr43vcasv0d5eq4',
          name: 'My Contact',
        },
      },
    })
    const { container } = renderComponent(storeWithContacts)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render empty state', () => {
    renderComponent(store)
    expect(screen.getByText('You have no contacts yet.')).toBeInTheDocument()
    expect(screen.getByText('Add contact')).toBeInTheDocument()
    expect(screen.queryByTestId('account-choice')).not.toBeInTheDocument()
  })

  it('should show Add Contact form overlay', async () => {
    renderComponent(store)
    await userEvent.click(screen.getByRole('button', { name: 'Add contact' }))
    expect(screen.getByPlaceholderText('Name')).toBeInTheDocument()
    expect(screen.getByPlaceholderText('Address')).toBeInTheDocument()
  })
})
