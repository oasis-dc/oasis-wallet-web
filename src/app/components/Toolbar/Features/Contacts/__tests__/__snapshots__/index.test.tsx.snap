// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Contacts  /> should match snapshot 1`] = `
.c0 {
  font-family: Rubik,sans-serif;
  font-size: 16px;
  line-height: 20px;
  background-color: #ffffff;
  color: #16213E;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  border: solid 1px #0500e2;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  padding: 12px;
  border-radius: 5px;
  cursor: pointer;
}

.c3:hover {
  background-color: rgba(183,183,183,0.5);
  color: black;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  margin-bottom: 48px;
  margin-left: -2px;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding-left: 2px;
  padding-right: 12px;
  overflow-y: auto;
}

.c11 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  height: 24px;
}

.c6 {
  font-size: 16px;
  line-height: 20px;
  font-weight: bold;
}

.c9 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  border: 2px solid #0500e2;
  border-radius: 18px;
  color: #16213E;
  padding: 4px 20px;
  font-size: 14px;
  line-height: 20px;
  background-color: #0500e2;
  color: #F8F8F8;
  border-radius: 18px;
  -webkit-transition-property: color,background-color,border-color,box-shadow;
  transition-property: color,background-color,border-color,box-shadow;
  -webkit-transition-duration: 0.1s;
  transition-duration: 0.1s;
  -webkit-transition-timing-function: ease-in-out;
  transition-timing-function: ease-in-out;
  font-weight: bold;
}

.c9:hover {
  background-color: rgba(183,183,183,0.4);
  color: #000000;
  box-shadow: 0px 0px 0px 2px #0500e2;
}

.c9:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #00A9FF;
}

.c9:focus > circle,
.c9:focus > ellipse,
.c9:focus > line,
.c9:focus > path,
.c9:focus > polygon,
.c9:focus > polyline,
.c9:focus > rect {
  outline: none;
  box-shadow: 0 0 2px 2px #00A9FF;
}

.c9:focus::-moz-focus-inner {
  border: 0;
}

.c9:focus:not(:focus-visible) {
  outline: none;
  box-shadow: none;
}

.c9:focus:not(:focus-visible) > circle,
.c9:focus:not(:focus-visible) > ellipse,
.c9:focus:not(:focus-visible) > line,
.c9:focus:not(:focus-visible) > path,
.c9:focus:not(:focus-visible) > polygon,
.c9:focus:not(:focus-visible) > polyline,
.c9:focus:not(:focus-visible) > rect {
  outline: none;
  box-shadow: none;
}

.c9:focus:not(:focus-visible)::-moz-focus-inner {
  border: 0;
}

.c13 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  border: 2px solid #0500e2;
  border-radius: 4px;
  color: #16213E;
  padding: 4px 22px;
  font-size: 16px;
  line-height: 20px;
  background-color: #0500e2;
  color: #F8F8F8;
  border-radius: 4px;
  -webkit-transition-property: color,background-color,border-color,box-shadow;
  transition-property: color,background-color,border-color,box-shadow;
  -webkit-transition-duration: 0.1s;
  transition-duration: 0.1s;
  -webkit-transition-timing-function: ease-in-out;
  transition-timing-function: ease-in-out;
  font-weight: bold;
}

.c13:hover {
  box-shadow: 0px 0px 0px 2px #0500e2;
}

.c13:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #00A9FF;
}

.c13:focus > circle,
.c13:focus > ellipse,
.c13:focus > line,
.c13:focus > path,
.c13:focus > polygon,
.c13:focus > polyline,
.c13:focus > rect {
  outline: none;
  box-shadow: 0 0 2px 2px #00A9FF;
}

.c13:focus::-moz-focus-inner {
  border: 0;
}

.c13:focus:not(:focus-visible) {
  outline: none;
  box-shadow: none;
}

.c13:focus:not(:focus-visible) > circle,
.c13:focus:not(:focus-visible) > ellipse,
.c13:focus:not(:focus-visible) > line,
.c13:focus:not(:focus-visible) > path,
.c13:focus:not(:focus-visible) > polygon,
.c13:focus:not(:focus-visible) > polyline,
.c13:focus:not(:focus-visible) > rect {
  outline: none;
  box-shadow: none;
}

.c13:focus:not(:focus-visible)::-moz-focus-inner {
  border: 0;
}

.c10 {
  border-radius: 4px;
}

@media only screen and (max-width:768px) {

}

@media only screen and (max-width:768px) {

}

@media only screen and (max-width:768px) {

}

@media only screen and (max-width:768px) {

}

@media only screen and (max-width:768px) {
  .c3 {
    border: solid 1px #0500e2;
  }
}

@media only screen and (max-width:768px) {
  .c3 {
    padding: 6px;
  }
}

@media only screen and (max-width:768px) {
  .c2 {
    margin-bottom: 24px;
  }
}

@media only screen and (max-width:768px) {
  .c2 {
    margin-left: -2px;
  }
}

@media only screen and (max-width:768px) {
  .c2 {
    padding-left: 2px;
  }
}

@media only screen and (max-width:768px) {
  .c2 {
    padding-right: 6px;
  }
}

@media only screen and (max-width:768px) {

}

@media only screen and (max-width:768px) {
  .c11 {
    height: 12px;
  }
}

<div
  class="c0"
  style="min-height: 100dvh;"
>
  <div
    class="c1"
  >
    <div
      class="c2"
      style="max-height: 400px;"
    >
      <div
        aria-checked="false"
        class="c3"
        data-testid="account-choice"
        role="checkbox"
        tabindex="0"
      >
        <div
          class="c4"
        >
          <div
            class="c5"
            data-testid="account-name"
          >
            <span
              class="c6"
            >
              My Contact
            </span>
          </div>
          <div
            class="c5"
          >
            <span
              class="c6"
            >
              <span
                class="notranslate"
                translate="no"
              >
                oasis1qq2v...sv0d5eq4
              </span>
            </span>
          </div>
          <div
            class="c7"
          >
            <div
              class="c8"
            >
              <button
                class="c9 c10"
                type="button"
              >
                Manage
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="c11"
    />
    <div
      class="c12"
    >
      <button
        class="c13"
        type="button"
      >
        Add contact
      </button>
    </div>
  </div>
</div>
`;
